File: CP_Sheet_PH/Loops/P_Minimize_Number.cpp
/*
________________________
|                      |
|     AB Atiq          |
|______________________|
*/

// Problem link: https://codeforces.com/group/MWSDmqGsZm/contest/219774/problem/P

#include <bits/stdc++.h>
using namespace std;

int main()
{
    int sz;
    cin >> sz;

    int arr[sz];

    for (int i = 0; i < sz; i++)
    {
        cin >> arr[i];
    }

    int cnt = 0;
    bool flag = true;

    while (flag)
    {
        int total_num_div = 0;

        for (int i = 0; i < sz; i++)
        {
            if (arr[i] % 2 == 0)
            {
                total_num_div++;
                arr[i] = arr[i] / 2;
            }
            else
            {
                flag = false;
                break;
            }
        }

        if (total_num_div == sz)
        {
            cnt++;
        }
    }

    cout << cnt << "\n";

    return 0;
}

// easy and sorter solution

#include <bits/stdc++.h>
using namespace std;
int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    cin >> n;
    int min_div = INT_MAX;
    for (int i = 0; i < n; i++)
    {
        int x;
        cin >> x;
        int cnt = 0;
        while (x % 2 == 0)
        {
            cnt++;
            x /= 2;
        }
        min_div = min(min_div, cnt);
    }
    cout << min_div << endl;
    return 0;
}

File: C-5-SDT/OOP_and_Python_Programming/M4_Ass_1/P_Minimize_Number.py
# https://codeforces.com/group/MWSDmqGsZm/contest/219774/problem/P
sz = int(input())
num_lst = [int(x) for x in input().split()]

cnt = 0
flag = True
while flag:
    total_num_div = 0
    for i, n in enumerate(num_lst):
        # pass
        if n%2==0:
            total_num_div+=1
            num_lst[i]=n/2
        else:
            break
    if total_num_div == sz:
        cnt += 1
    else:
        flag = False
        
print(cnt)

# give me solution without - for i, n in enumerate(num_lst):
sz = int(input())
num_lst = [int(x) for x in input().split()]

cnt = 0
flag = True
while flag:
    total_num_div = 0
    i = 0
    while i < sz:
        n = num_lst[i]
        # pass
        if n%2==0:
            total_num_div+=1
            num_lst[i]=n/2
        else:
            break
        i += 1
    if total_num_div == sz:
        cnt += 1
    else:
        flag = False
        
print(cnt)


# optimized solution
n = int(input())
arr = list(map(int,input().split()))

operation = 0
flag = True
while flag:
    num_of_div = 0
    for i in range(0,n):
        if arr[i] % 2 == 0:
            arr[i] = arr[i] / 2 
            num_of_div = num_of_div + 1
        else:
            break
    if num_of_div == n:
        operation += 1
    else:
        flag = False

print(operation)